runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala

```diff
@@ -0,0 +1,114 @@
+
+    def encode(value: String): String = {
+      val valueBytes = value.getBytes(StandardCharsets.UTF_8)
+      val firstIndexToEscape = valueBytes.indexWhere(isEscapingChar(_))
+      if (firstIndexToEscape == -1) {
+        // Why not return original input?
+        // The following is not always true for any value: `new String(value.getBytes(StandardCharsets.UTF_8)) == value`
+        // Consider "a \ud801": its underlying array of bytes is [97, 0, 32, 0, 1, -40] while "a \ud801".getBytes(StandardCharsets.UTF_8) yields [97, 32, 63]
+        // There are 2 effects of decoding: ascii characters are encoded with 1 byte and lone surrogate has been replaced with ? (63 in ascii)
+        new String(valueBytes, StandardCharsets.US_ASCII)
+      } else
+        encodeSlow(valueBytes, firstIndexToEscape)
+    }
+
+    private def isEscapingChar(b: Byte): Boolean = b < ' ' || b >= '~' || b == '%'
+
+    private def encodeSlow(valueBytes: Array[Byte], riArg: Int): String = {
+      var ri = riArg
+      val escapedBytes = new Array[Byte](riArg + ((valueBytes.length - riArg) * 3))
+      // copy over the good bytes
+      if (riArg != 0) System.arraycopy(valueBytes, 0, escapedBytes, 0, riArg)
+      var wi = ri
+
+      while (ri < valueBytes.length) {
+        val b = valueBytes(ri)
+        // Manually implement URL encoding, per the gRPC spec.
+        if (isEscapingChar(b)) {
+          escapedBytes.update(wi, '%')
+          escapedBytes.update(wi + 1, HexArr((b >> 4) & 0xf))
+          escapedBytes.update(wi + 2, HexArr(b & 0xf))
+          wi += 3
+        } else {
+          escapedBytes.update(wi, b)
+          wi += 1
+        }
+        ri += 1
+      }
+      new String(escapedBytes, 0, wi, StandardCharsets.US_ASCII)
+    }
+  }
+
+  // Copied with slight adaptations from https://github.com/grpc/grpc-java/blob/79e75bace40cea7e4be72e7dcd1f41c3ad6ee857/api/src/main/java/io/grpc/Status.java#L626
+  object Decoder {
+    private val TransferEncoding = StandardCharsets.US_ASCII
+
+    def decode(value: String): String =
+      if (value.indexOf('%') > -1)
+        decodeSlow(value)
+      else
+        value
+
+    private def decodeSlow(value: String): String = {
+      val source = value.getBytes(TransferEncoding)
+      val buf = ByteBuffer.allocate(source.length)
+      var i = 0
+      while (i < source.length) {
+        if (source(i) == '%' && i + 2 < source.length) {
+          val ch0 = Character.digit(source(i + 1), 16)
+          val ch1 = Character.digit(source(i + 2), 16)
+          if (ch0 > -1 && ch1 > -1) {
+            val res = (ch0 << 4) + ch1
+            buf.put(res.toByte)
+            i += 3
+          } else {
+            buf.put(source(i))
+            i += 1
+          }
+        } else {
+          buf.put(source(i))
+          i += 1
+        }
+      }
+      new String(buf.array, 0, buf.position(), StandardCharsets.UTF_8)
+    }
+  }
```

Comment:
It would be interesting to bench the solution above against the following:

```scala
private[this] final def decodeSlow(value: String): String = {
  @tailrec def loop(buf: ByteBuffer, source: Array[Byte], at: Int): String =
    if (at < source.length) {
      val cur = source(at)
      val skip = 
        if (cur != '%' || (at + 2) > source.length) {
          buf.put(cur)
          1
        } else {
          val ch0 = Character.digit(source(at + 1), 16)
          val ch1 = Character.digit(source(at + 2), 16)
          if (ch0 > -1 && ch1 > -1) {
            buf.put(((ch0 << 4) + ch1).toByte)
            3
          } else {
            buf.put(cur)
            1
          }
        }
      loop(buf, source, at + skip)
    } else new String(buf.array, 0, buf.position, StandardCharsets.UTF_8)

  val source = value.getBytes(TransferEncoding)
  val buf = ByteBuffer.allocate(value.length)

  loop(buf, source, 0)
}
```

Another possible improvement is to pass in the `idx` found by `indexOf('%')` and do an arraycopy from 0 … idx to avoid having to put those chars 1-by-1, as the worst-case the '%' is at the end of the string. Then you can start the `loop` at `idx` and progress from there.

Reply:
It's good that we try to find a decent performance level for this, but note that it's only for errors and that shouldn't be extremely performance sensitive.

Reply:
With the code pasted by @viktorklang I got `0,652 ± 0,003  us/op` while with the code currently in the PR I got `0,609 ± 0,009  us/op`.

I agree with @patriknw that it's good to have reasonable performance but IMO it's not critical to get the best since it's only for errors